File: solution/leetcode/104.java
// https://leetcode.com/problems/maximum-depth-of-binary-tree/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
	private int _maxDepth;
	
	private void recur(TreeNode root, int curDepth) {
		if (root == null)
			return;
		_maxDepth = Math.max(_maxDepth, curDepth);
		recur(root.left, curDepth + 1);
		recur(root.right, curDepth + 1);
	}
	
    public int maxDepth(TreeNode root) {
        _maxDepth = 0;
        recur(root, 1);
        return _maxDepth;
    }
}

File: solution/leetcode/104.py
# https://leetcode.com/problems/maximum-depth-of-binary-tree/submissions/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        if root == None:
            return 0
        return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right)) 

'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # help from discussion
    # A BFS solution
    def maxDepth(self, root: TreeNode) -> int:
        if root == None:
            return 0
        from collections import deque
        stack = deque([(root, 1)])
        while stack:
            var, depth = stack.popleft()
            if not stack and not var.left and not var.right:
                return depth
            if var.left:
                stack.append((var.left, depth+1))
            if var.right:
                stack.append((var.right, depth+1))
'''
